Snooze alarm until you re-enable and improvements in snoozeactivity

JohanDegraeve opened the pull request:
- possibility to snooze 10 years, this corresponds to for-ever. The
  corresponding text in the snoozeValue list will be "until you
  re-enable".
- Added message in snoozeactivity indicating if all, low and/or high
  alerts are disabled and for how long
- text is changed when re-enabling alerts
- all to displaystatus in a separate thread, to make sure new texts are
  always correctly shown, override onResume, also make call to
  displaystatus

JohanDegraeve committed:
Possibility to snooze alarm until re-enabled  and improvements in text
shown in SnoozeActivity
- possibility to snooze 10 years, this corresponds to for-ever. The
  corresponding text in the snoozeValue list will be "until you
  re-enable".
- Added message in snoozeactivity indicating if all, low and/or high
  alerts are disabled and for how long
- text is changed when re-enabling alerts
- all to displaystatus in a separate thread, to make sure new texts are
  always correctly shown, override onResume, also make call to
  displaystatus

JohanDegraeve committed:
Possibility to snooze alarm until re-enabled  and improvements in text
shown in SnoozeActivity
- possibility to snooze 10 years, this corresponds to for-ever. The
  corresponding text in the snoozeValue list will be "until you
  re-enable".
- Added message in snoozeactivity indicating if all, low and/or high
  alerts are disabled and for how long
- text is changed when re-enabling alerts
- all to displaystatus in a separate thread, to make sure new texts are
  always correctly shown, override onResume, also make call to
  displaystatus

JohanDegraeve commented:
by teh way, there's still one thing I couldn't solve : 
If you have the snoozeactivity open, no active alerts, no alearts disabled. Just the text "No active alert exists" and the three buttons (disable low, high, all alerts).
Keep the app open on that screen.
Then there's an alert, a notification appears on top, scrol it down, select it. The snoozeativity opens, but the buttonSnooze and snoozeValue are not popping up. Something to do with the fact that the visibility is set to View.GONE

AdrianLxM commented:
Not even overriding `protected void onNewIntent(Intent intent)` worked for me.
What should work however is overriding `onWindowFocusChanged(boolean hasFocus)`:

``` java
    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus){
            displayStatus();        
        }
    }
```

AdrianLxM commented on app/src/main/java/com/eveningoutpost/dexdrip/SnoozeActivity.java:
```diff
@@ -293,24 +299,57 @@ void displayStatus() {
+                status += "\n\nHigh alerts disabled until " + textToAdd;
+            }
+        }
+        alertStatus.post(new Runnable() {
```
Is there any reason to post that to the element.
We are on an UI thread anyways. This could lead to lags in updating the UI (seen with the Database-Import that also posts to the UI).

tzachi-dar replied:
Agree with adrian, this can be called directly.

Commit:
SnoozeActivity-updates

- direct setText - no delay
- displayStatus when activity gets focus again (e.g. after opening the notification area to click on an alarm)
- display "rerise" as it is, not as snooze
- danging else
- some indentation changes my IDE made automatically

Commit:
SnoozeActivity-updates

- direct setText - no delay
- displayStatus when activity gets focus again (e.g. after opening the notification area to click on an alarm)
- display "rerise" as it is, not as snooze
- danging else
- some indentation changes my IDE made automatically

AdrianLxM commented:
- I tested it (with "onWindowFocusChanged" and direct setText) and it works fine.
- I'm good with 10 years being set as "a very long time to snooze" and "at least one year from now" to check for it.
- An alarm that is in its "rerise" period still shows as snoozed.

I sent you a PR with my version that I tested and works for me: https://github.com/JohanDegraeve/xDrip-Experimental/pull/3

There are still some issues:
- Should we give the user a Button "re-enable" alert? For us it seems logical to simply disable/reenable all alerts.
- After snoozing "forever", the status shows "Alert snoozed until 21:14:21 (30456 minutes left)". Couldn't we change that so something more meaningful. (The 30456 isn't right anyways). This happens with your (original) version and with my update.

tzachi-dar commented on app/src/main/java/com/eveningoutpost/dexdrip/SnoozeActivity.java:
```diff
@@ -293,24 +299,57 @@ void displayStatus() {
+            if (prefs.getLong("high_alerts_disabled_until", 0) > now) {
+                String textToAdd = (prefs.getLong("high_alerts_disabled_until", 0) > now + (infiniteSnoozeValueInMinutes - 365 * 24 * 60) * 60 * 1000)
+                        //if high alerts would have been disabled "until you re-enable", and this test is done less than 365 * 24 * 60 minutes later, then this test will give true
+                        ? "you re-enable":DateFormat.getTimeInstance(DateFormat.MEDIUM).format(new Date(prefs.getLong("high_alerts_disabled_until", 0)));
```
you have here the code that does the calculation of how long the alerts are disabled, 3 times.
Please move that to a single static function and call that also from the home activity that also displays, low alerts disabled...

JohanDegraeve replied:
i'm not sure if this is worth putting in a function. It's just a single line of code, and in Home.java the check is slightly different

AdrianLxM replied:
@tzachi-dar, are you thinking of a function `private String timeDisabledUntil(SharedPreferences prefs, String lookupString, long now)`? Seems like a parameter list almost as long as the code itself ;)

tzachi-dar commented:
Disabling for ever, can make a reasonable choice for all alerts or low/high, but I don't think that it is reasonable for a specific alert.
It seems to me that we should allow this option only for these 3 cases, and if we want to disable an alert for a long period we should add an option to enable/disable each alert alone. This way the UI will show which alerts are on and which are not.

JohanDegraeve committed:
Merge pull request #3 from AdrianLxM/patch-1

SnoozeActivity-updates

JohanDegraeve commented:
I will still test and see if I can include the comments

JohanDegraeve committed:
disable for ever only for disable all high/low or all - not for specific bg alerts

Commit:
Merge pull request #217 from StephenBlackWasAlreadyTaken/bgreadings-shortcut

Bgreadings shortcut

AdrianLxM commented:
> Disabling for ever, can make a reasonable choice for all alerts or low/high, but I don't think that it is reasonable for a specific alert.
> It seems to me that we should allow this option only for these 3 cases, and if we want to disable an alert for a long period we should add an option to enable/disable each alert alone. This way the UI will show which alerts are on and which are not.

There is a difference between disabling an alert and long-snoozing it. No matter how long you snooze an alert, if any other alert with higher priority will become active, this alert will get cancelled and therefore the snooze time is automatically over.

I don't see that many usecases for me for that either. But it fits right into the existing logic so adding one more time won't hurt.

JohanDegraeve commented:
I'm restricting the disable for ever for the disable all only. It's true it doesn't make much sense to do this for active bg alerts.
There's two users here who use the nightwatch app to monitor their kids. They prefer to be able to disable the alerts on the xdrip app, but still in exceptional cases (like during holidays) they want to use xdrip alarms, in that case they will re-enable the all's

I'm testing

tzachi-dar commented:
Please note that to disable bg alerts, one can also change profile to
silence... But i'm ok with other way to silence them.
בתאריך 13 בדצמ 2015 18:02,‏ "Johan Degraeve" notifications@github.com כתב:

> I'm restricting the disable for ever for the disable all only. It's true
> it doesn't make much sense to do this for active bg alerts.
> There's two users here who use the nightwatch app to monitor their kids.
> They prefer to be able to disable the alerts on the xdrip app, but still in
> exceptional cases (like during holidays) they want to use xdrip alarms, in
> that case they will re-enable the all's
> 
> I'm testing
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/215#issuecomment-164270917
> .

JohanDegraeve committed:
conflict fix

JohanDegraeve committed:
no need to handle infiniteSnoozeValueInMinutes in getNameFromTime

JohanDegraeve commented:
disable "until you re-enable" is now only for the disable all buttons

please check

tzachi-dar commented:
ok, played with it a bit, looks nice.
Thank you.
